File: backend/app/schemas/topology.py
from pydantic import BaseModel, Field
from typing import List, Dict, Any


class TopologyNode(BaseModel):
    """Node representing a network segment in topology graph"""
    id: str
    label: str
    ip_range: str
    zone_type: str
    color: str
    description: str | None = None


class TopologyEdge(BaseModel):
    """Edge representing firewall rule(s) between segments"""
    id: str
    from_: str = Field(..., alias='from', serialization_alias='from')  # source segment id
    to: str  # destination segment id
    label: str
    metadata: Dict[str, Any] = {}

    class Config:
        populate_by_name = True


class TopologyGraph(BaseModel):
    """Complete topology graph with nodes and edges"""
    nodes: List[TopologyNode]
    edges: List[TopologyEdge]


class PathSegment(BaseModel):
    segment_id: int
    segment_name: str


class RuleApplied(BaseModel):
    rule_id: int
    rule_name: str
    protocol: str
    port_range: str | None = None


class PathAnalysisRequest(BaseModel):
    source_segment_id: int
    destination_segment_id: int
    protocol: str | None = None
    port: int | None = None


class PathAnalysisResponse(BaseModel):
    reachable: bool
    path: List[PathSegment]
    rules_applied: List[RuleApplied]


class AffectedConnection(BaseModel):
    source_segment: str
    destination_segment: str
    protocol: str
    port: str | None = None


class DependentPath(BaseModel):
    source: str
    destination: str
    via: List[str]


class RuleImpactRequest(BaseModel):
    rule_id: int


class RuleImpactResponse(BaseModel):
    affected_connections: List[AffectedConnection]
    dependent_paths: List[DependentPath]
    warning: str

File: frontend/src/types/topology.ts
export interface TopologyNode {
  id: string;
  label: string;
  ip_range: string;
  zone_type: 'DMZ' | 'Internal' | 'External' | 'Management';
  color: string;
  description?: string;
}

export interface TopologyEdge {
  id: string;
  from: string;
  to: string;
  label: string;
  metadata: {
    rule_ids: number[];
    protocols: string[];
    ports: string[];
    actions: string[];
    descriptions: string[];
  };
}

export interface TopologyGraph {
  nodes: TopologyNode[];
  edges: TopologyEdge[];
}

export interface NetworkSegment {
  id: number;
  name: string;
  ip_range: string;
  zone_type: string;
  color: string;
  description?: string;
  created_at: string;
  updated_at: string;
}

export interface Firewall {
  id: number;
  name: string;
  vendor?: string;
  model?: string;
  management_ip?: string;
  created_at: string;
  updated_at: string;
}

export interface FirewallRule {
  id: number;
  firewall_id: number;
  rule_name: string;
  source_segment_id: number;
  destination_segment_id: number;
  protocol: string;
  port_range?: string;
  action: 'ALLOW' | 'DENY';
  description?: string;
  created_at: string;
  updated_at: string;
}

export interface PathSegment {
  segment_id: number;
  segment_name: string;
}

export interface RuleApplied {
  rule_id: number;
  rule_name: string;
  protocol: string;
  port_range?: string;
}

export interface PathAnalysisResult {
  reachable: boolean;
  path: PathSegment[];
  rules_applied: RuleApplied[];
}

export interface AffectedConnection {
  source_segment: string;
  destination_segment: string;
  protocol: string;
  port?: string;
}

export interface DependentPath {
  source: string;
  destination: string;
  via: string[];
}

export interface RuleImpactResult {
  affected_connections: AffectedConnection[];
  dependent_paths: DependentPath[];
  warning: string;
}

export interface SearchResult {
  segments: Array<{
    id: number;
    name: string;
    ip_range: string;
    zone_type: string;
  }>;
  rules: Array<{
    id: number;
    rule_name: string;
    protocol: string;
    action: string;
  }>;
}
